File: plugins/commands/serve/client/communicator.rb
# Copyright (c) HashiCorp, Inc.
# SPDX-License-Identifier: MIT


module VagrantPlugins
  module CommandServe
    class Client
      class Communicator < Client
        # Generate callback and spec for required arguments
        #
        # @return [SDK::FuncSpec, Proc]
        def ready_func
          spec = client.ready_spec(Empty.new)
          cb = proc do |args|
            client.ready(args).ready
          end
          [spec, cb]
        end

        # @param [Vagrant::Machine]
        # @return [bool]
        def ready(machine)
          run_func(machine)
        end

        # Generate callback and spec for required arguments
        #
        # @return [SDK::FuncSpec, Proc]
        def wait_for_ready_func
          spec = client.wait_for_ready_spec(Empty.new)
          cb = proc do |args|
            client.wait_for_ready(args).ready
          end
          [spec, cb]
        end

        # @param [Vagrant::Machine]
        # @param [Integer] duration Timeout in seconds.
        # @return [Boolean]
        def wait_for_ready(machine, time)
          run_func(machine, Type::Duration.new(value: time))
        end

        # Generate callback and spec for required arguments
        #
        # @return [SDK::FuncSpec, Proc]
        def download_func
          spec = client.download_spec(Empty.new)
          cb = proc do |args|
            client.download(args)
          end
          [spec, cb]
        end

        # @param [Vagrant::Machine]
        # @param [String] remote path
        # @param [String] local path
        def download(machine, from, to)
          from = Pathname.new(from.to_s) if !from.is_a?(Pathname)
          to = Pathname.new(to.to_s) if !to.is_a?(Pathname)

          run_func(
            Type::NamedArgument.new(name: "to", value: to),
            Type::NamedArgument.new(name: "from", value: from),
            machine
          )
        end

        # Generate callback and spec for required arguments
        #
        # @return [SDK::FuncSpec, Proc]
        def upload_func
          spec = client.upload_spec(Empty.new)
          cb = proc do |args|
            client.upload(args)
          end
          [spec, cb]
        end

        # @param [Vagrant::Machine]
        # @param [String] local path
        # @param [String] remote path
        def upload(machine, from, to)
          from = Pathname.new(from.to_s) if !from.is_a?(Pathname)
          to = Pathname.new(to.to_s) if !to.is_a?(Pathname)

          run_func(
            Type::NamedArgument.new(name: "source", value: from),
            Type::NamedArgument.new(name: "destination", value: to),
            machine
          )
        end

        # Generate callback and spec for required arguments
        #
        # @return [SDK::FuncSpec, Proc]
        def execute_func
          spec = client.execute_spec(Empty.new)
          cb = proc do |args|
            client.execute(args)
          end
          [spec, cb]
        end

        # @param [Vagrant::Machine]
        # @param [String] command to run
        # @param [Hash] options
        # @return [Integer]
        def execute(machine, cmd, opts)
          opts = {} if opts.nil?
          run_func(machine,
            Type::Options.new(value: opts),
            Type::CommunicatorCommandArguments.new(value: cmd)
          )
        end

        # Generate callback and spec for required arguments
        #
        # @return [SDK::FuncSpec, Proc]
        def privileged_execute_func
          spec = client.privileged_execute_spec(Empty.new)
          cb = proc do |args|
            client.privileged_execute(args)
          end
          [spec, cb]
        end

        # @param [Vagrant::Machine]
        # @param [String] command to run
        # @param [Hash] options
        # @return [Integer]
        def privileged_execute(machine, cmd, opts)
          opts = {} if opts.nil?
          run_func(machine,
            Type::Options.new(value: opts),
            Type::CommunicatorCommandArguments.new(value: cmd)
          )
        end

        # Generate callback and spec for required arguments
        #
        # @return [SDK::FuncSpec, Proc]
        def test_func
          spec = client.test_spec(Empty.new)
          cb = proc do |args|
            client.test(args).valid
          end
          [spec, cb]
        end

        # @param [Vagrant::Machine]
        # @param [String] command to run
        # @param [Hash] options
        # @return [Boolean]
        def test(machine, cmd, opts)
          opts = {} if opts.nil?
          run_func(machine,
            Type::Options.new(value: opts),
            Type::CommunicatorCommandArguments.new(value: cmd)
          )
        end

        # Generate callback and spec for required arguments
        #
        # @return [SDK::FuncSpec, Proc]
        def reset_func
          spec = client.reset_spec(Empty.new)
          cb = proc do |args|
            client.reset(args)
          end
          [spec, cb]
        end

        # Reset the communicator connection
        #
        # @param machine [Vagrant::Machine] Guest to reset connection on
        def reset(machine)
          run_func(machine)
        end
      end
    end
  end
end

File: builtin/myplugin/communicator/communicator.go
// Copyright (c) HashiCorp, Inc.
// SPDX-License-Identifier: MIT

package communicator

import (
	"github.com/hashicorp/go-argmapper"
	"github.com/hashicorp/go-hclog"
	"github.com/hashicorp/vagrant-plugin-sdk/component"
	plugincore "github.com/hashicorp/vagrant-plugin-sdk/core"
	pb "github.com/hashicorp/vagrant/builtin/myplugin/proto"
)

type DummyConfig struct {
}

// DummyCommunicator is a Communicator implementation for myplugin.
type DummyCommunicator struct {
	config DummyConfig
}

func (h *DummyCommunicator) MatchFunc() interface{} {
	return h.Match
}

func (h *DummyCommunicator) Match(machine plugincore.Machine) (isMatch bool, err error) {
	return true, nil
}

func (h *DummyCommunicator) InitFunc() interface{} {
	return h.Init
}

func (h *DummyCommunicator) Init(machine plugincore.Machine) error {
	return nil
}

func (h *DummyCommunicator) ReadyFunc() interface{} {
	return h.Ready
}

func (h *DummyCommunicator) Ready(machine plugincore.Machine) (isReady bool, err error) {
	return false, nil
}

func (h *DummyCommunicator) WaitForReadyFunc() interface{} {
	return h.WaitForReady
}

func (h *DummyCommunicator) WaitForReady(machine plugincore.Machine, wait int) (isReady bool, err error) {
	return false, nil
}

func (h *DummyCommunicator) DownloadFunc() interface{} {
	return h.Download
}

func (h *DummyCommunicator) Download(input struct {
	argmapper.Struct
	Machine     plugincore.Machine `argmapper:",typeOnly"`
	Logger      hclog.Logger       `argmapper:",typeOnly"`
	Source      string
	Destination string
},
) error {
	input.Logger.Debug("got Source ", input.Source)
	input.Logger.Debug("got Destination ", input.Destination)
	return nil
}

func (h *DummyCommunicator) UploadFunc() interface{} {
	return h.Upload
}

func (h *DummyCommunicator) Upload(input struct {
	argmapper.Struct
	Machine     plugincore.Machine `argmapper:",typeOnly"`
	Logger      hclog.Logger       `argmapper:",typeOnly"`
	Source      string
	Destination string
},
) error {
	input.Logger.Debug("got Source ", input.Source)
	input.Logger.Debug("got Destination ", input.Destination)
	return nil
}

func (h *DummyCommunicator) ExecuteFunc() interface{} {
	return h.Execute
}

func (h *DummyCommunicator) Execute(
	machine plugincore.Machine,
	command []string,
	options *pb.CommunicatorOptions,
) (status int32, err error) {
	return 0, nil
}

func (h *DummyCommunicator) PrivilegedExecuteFunc() interface{} {
	return h.PrivilegedExecute
}

func (h *DummyCommunicator) PrivilegedExecute(
	machine plugincore.Machine,
	command []string,
	options *pb.CommunicatorOptions,
) (status int32, err error) {
	return 0, nil
}

func (h *DummyCommunicator) TestFunc() interface{} {
	return h.Test
}

func (h *DummyCommunicator) Test(
	machine plugincore.Machine,
	command []string,
	options ...pb.CommunicatorOptions,
) (valid bool, err error) {
	return true, nil
}

func (h *DummyCommunicator) ResetFunc() interface{} {
	return h.Reset
}

func (h *DummyCommunicator) Reset(machine plugincore.Machine) (err error) {
	return nil
}

var (
	_ component.Communicator = (*DummyCommunicator)(nil)
)
